File: 백준/Gold/11779. 최소비용 구하기 2/최소비용 구하기 2.js
class MinHeap {
  constructor() {
    this.heap = [];
  }

  push(node) {
    this.heap.push(node);
    this.heapifyUp();
  }

  pop() {
    if (this.heap.length === 1) return this.heap.pop();
    const min = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.heapifyDown();
    return min;
  }

  swap(i, j) {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
  }

  heapifyUp() {
    let idx = this.heap.length - 1;
    while (idx > 0) {
      let parentIdx = Math.floor((idx - 1) / 2);
      if (this.heap[parentIdx][1] <= this.heap[idx][1]) break;
      this.swap(parentIdx, idx);
      idx = parentIdx;
    }
  }

  heapifyDown() {
    let idx = 0;
    while (true) {
      let leftChildIdx = 2 * idx + 1;
      if (leftChildIdx >= this.heap.length) break;

      let rightChildIdx = 2 * idx + 2;
      let smallerChild = leftChildIdx;

      if (
        rightChildIdx < this.heap.length &&
        this.heap[rightChildIdx][1] < this.heap[smallerChild][1]
      )
        smallerChild = rightChildIdx;

      if (this.heap[idx][1] <= this.heap[smallerChild][1]) break;

      this.swap(idx, smallerChild);
      idx = smallerChild;
    }
  }

  isEmpty() {
    return this.heap.length === 0;
  }
}

const fs = require('fs');
const filePath = process.platform === 'linux' ? 0 : 'input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const n = Number(input[0]);
const m = Number(input[1]);

const graph = Array.from({ length: n + 1 }, () => []);
for (let i = 2; i < m + 2; i++) {
  const [from, to, cost] = input[i].split(' ').map(Number);
  graph[from].push([to, cost]);
}

const [start, end] = input[input.length - 1].split(' ').map(Number);
dijkstra(start, end);

function dijkstra(start, end) {
  const dist = Array(n + 1).fill(100_000_000);
  const pq = new MinHeap();

  dist[start] = 0;
  pq.push([start, 0]);

  const prev = Array(n + 1);

  while (!pq.isEmpty()) {
    const [curNode, curDist] = pq.pop();

    if (curNode === end) break;
    if (dist[curNode] < curDist) continue;

    for (let [adj, cost] of graph[curNode]) {
      const newDist = curDist + cost;
      if (dist[adj] > newDist) {
        dist[adj] = newDist;
        pq.push([adj, newDist]);
        prev[adj] = curNode;
      }
    }
  }

  const path = [];
  let last = end;
  while (last !== start) {
    path.push(last);
    last = prev[last];
  }
  path.push(start);

  const result = [dist[end], path.length, path.reverse().join(' ')];
  console.log(result.join('\n'));
}

File: 백준/Gold/11779. 최소비용 구하기 2/최소비용 구하기 2.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static class Bus implements Comparable<Bus> {
        int to, cost;

        public Bus(int to, int cost) {
            this.to = to;
            this.cost = cost;
        }

        public int compareTo(Bus o) {
            return this.cost - o.cost;
        }
    }

    static int n;
    static ArrayList<Bus>[] city;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        city = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            city[i] = new ArrayList<>();
        }

        StringTokenizer st = null;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());

            city[from].add(new Bus(to, cost));
        }

        st = new StringTokenizer(br.readLine(), " ");
        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());

        System.out.println(dijkstra(start, end));
    }

    static String dijkstra(int start, int end) {
        int[] dist = new int[n + 1];
        Arrays.fill(dist, 100_000_000);
        dist[start] = 0;
        PriorityQueue<Bus> pq = new PriorityQueue<>();
        pq.add(new Bus(start, 0));

        int[] prev = new int[n + 1]; // 최단 경로로 왔을 때 직전 도시

        while (!pq.isEmpty()) {
            Bus cur = pq.poll();
            if (cur.to == end) break;

            if (cur.cost > dist[cur.to]) continue;

            for (Bus bus : city[cur.to]) {
                if (dist[bus.to] > dist[cur.to] + bus.cost) {
                    dist[bus.to] = dist[cur.to] + bus.cost;
                    pq.add(new Bus(bus.to, dist[bus.to]));
                    prev[bus.to] = cur.to;
                }
            }
        }

        ArrayList<Integer> path = new ArrayList<>();
        int last = end;
        while (last != start) {
            path.add(last);
            last = prev[last];
        }
        path.add(start);

        StringBuilder sb = new StringBuilder();
        sb.append(dist[end]).append("\n").append(path.size()).append("\n");
        for (int i = path.size() -1; i >= 0; i--) {
            sb.append(path.get(i)).append(" ");
        }

        return sb.toString();
    }
}